File: SwordDemoLibrary/pluginable/src/main/java/com/sword/pluginable/MainActivity.java
package com.sword.pluginable;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.example.utilclass.TextUtil;
import com.example.utilclass.ToastUtilKt;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import dalvik.system.DexClassLoader;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okio.BufferedSink;
import okio.Okio;
import okio.Source;

public class MainActivity extends AppCompatActivity {
	private static final String HOTFIX_DEX_NAME = "hotfix.dex";
	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		//reflectInvokePrivate();
		//refectInvokeOtherApk();
	}

	/**
	 * 反射访问 public 修饰的类
	 */
	private void reflectInvoke() {
		Class<?> cls = PublicUtils.class;
		try {
			Object publicUtils = cls.newInstance();
			Method shoutMethod = cls.getDeclaredMethod("shout");
			shoutMethod.invoke(publicUtils);
		} catch (IllegalAccessException | InstantiationException | NoSuchMethodException | InvocationTargetException e) {
			e.printStackTrace();
		}

	}


	/**
	 * 反射访问非公开的类，通过 class.forName("${className}")
	 */
	private void reflectInvokePrivate() {
		try {
			//执行 PrivateUtils#shout()
			Class<?> cls = Class.forName("com.sword.pluginable.PrivateUtils");
			//获取构造方法，并通过构造方法获取实例
			Constructor<?> constructor = cls.getDeclaredConstructors()[0];
			constructor.setAccessible(true);
			Object PrivateUtils = constructor.newInstance();
			//获取 PrivateUtils#shout 方法实例
			Method shoutMethod = cls.getDeclaredMethod("shout");
			shoutMethod.setAccessible(true);
			//执行 PrivateUtils.shout() 方法
			shoutMethod.invoke(PrivateUtils);
		} catch (ClassNotFoundException | IllegalAccessException | InstantiationException | NoSuchMethodException | InvocationTargetException e) {
			e.printStackTrace();
		}
	}

	/**
	 * 访问同一个工程下其他 apk 模块的类，使用 ClassLoader
	 */
	private void refectInvokeOtherApk() {
		//先将 assets/plugin-debug.apk 拷贝到缓存目录中
		File apk = new File(getCacheDir() + "/plugin.apk");
		if (!apk.exists()) {
			try(Source source = Okio.source(getAssets().open("plugin-debug.apk"));
			    BufferedSink sink = Okio.buffer(Okio.sink(apk))) {
				sink.writeAll(source);
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
		if (apk.exists()) {
			//创建插件 apk 对应的 ClassLoader，这个 ClassLoader 加载插件 apk 中所有的类。
			DexClassLoader classLoader = new DexClassLoader(apk.getPath(), getCacheDir().getPath(), null, null);
			try {
				Class<?> cls = classLoader.loadClass("com.example.plugin.PublicUtils");
				Object publicUtils = cls.newInstance();
				Method shoutMethod = cls.getDeclaredMethod("shout");
				shoutMethod.invoke(publicUtils);
			} catch (ClassNotFoundException | IllegalAccessException | InstantiationException | NoSuchMethodException | InvocationTargetException e) {
				e.printStackTrace();
			}
		}

	}

	public void onClick(View view) {
		if (view.getId() == R.id.loadPlugin) {
			refectInvokeOtherApk();
		}
		if (view.getId() == R.id.loadHofixApk) {
			downloadHotfix();
			copyAssetsFileToCache("hotfix.apk", "hotfix.apk");
		}
		if (view.getId() == R.id.loadHofixDex) {
			copyAssetsFileToCache("hotfix.dex", "hotfix.dex");
		}
		if (view.getId() == R.id.showTitle) {
			Title title = new Title();
			((TextView)findViewById(R.id.title)).setText(title.getTitle());
		}

		if (view.getId() == R.id.removeHotfix) {
			File hotfixDex = new File(getCacheDir() + File.separator + "hotfix.dex");
			if (hotfixDex.exists() && hotfixDex.delete()) {
				Toast.makeText(this, "unstall hotfix success, please restart app", Toast.LENGTH_SHORT).show();
			}
		}

		if (view.getId() == R.id.restart) {
			Intent intent = new Intent(this, MainActivity.class);
			intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
			startActivity(intent);
			finish();
			Runtime.getRuntime().exit(0);
		}

	}

	private void downloadHotfix() {
		OkHttpClient client = new OkHttpClient();
		final Request request = new Request.Builder()
				.url("https://api.hencoder.com/patch/upload/hotfix.dex")
				.build();

		client.newCall(request).enqueue(new Callback() {
			@Override
			public void onFailure(@NonNull Call call, @NonNull IOException e) {
				runOnUiThread(() -> {
					ToastUtilKt.toast(MainActivity.this, "补丁下载失败");
				});
			}

			@Override
			public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
				try(BufferedSink sink = Okio.buffer(Okio.sink(new File(getCacheDir() + File.separator + HOTFIX_DEX_NAME)))) {
					sink.write(response.body().bytes());
				} catch (Exception e) {
					e.printStackTrace();
				}
				ToastUtilKt.toast(MainActivity.this, "补丁加载成功");
			}
		});
	}

	private void copyAssetsFileToCache(String filePathInAssets, String filePathInCache) {
		//拷贝热修复 apk 到缓存目录
		File apk = new File(getCacheDir() + File.separator + filePathInCache);
		try(Source source = Okio.source(getAssets().open(filePathInAssets));
				BufferedSink sink = Okio.buffer(Okio.sink(apk))) {
			sink.writeAll(source);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: SwordDemoLibrary/learnkotlin/src/main/java/com/example/learnkotlin/main/MainActivity.kt
package com.example.learnkotlin.main

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity
import com.example.learnkotlin.R
import com.example.learnkotlin.entity.User
import com.example.learnkotlin.lesson.LessonActivity
import com.example.learnkotlin.main.widget.CodeView
import com.example.learnkotlin.utils.Utils.toast
import com.example.learnkotlin.utils.get
import com.example.learnkotlin.utils.save

class MainActivity : AppCompatActivity(), View.OnClickListener {
    private val usernameKey = "username"
    private val passwordKey = "password"
    
    private lateinit var et_username: EditText
    private lateinit var et_password: EditText
    private lateinit var et_code: EditText
    
    @SuppressLint("CutPasteId")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        
        et_username = findViewById(R.id.et_username)
        et_password = findViewById(R.id.et_username)
        et_code = findViewById(R.id.et_code)
        
        
        
        //eCode 是 Button! 类型，表示这个类型不是 kotlin 声明出来的类型，而是别的平台声明出来的类型，kotlin 不能判断这个类型是否为空，因此只能用平台类型来接收它。平台类型泛指来自其他平台的类型，因为 kotlin 可以跟多种语言做交互。
        //编写代码的时候，编译器会把平台类型看成不可空类型。如果平台类型的对象为空，在使用该对象的时候依然会报空指针，可以看出使用 kotlin 并不代表永远的避免了空指针异常
        //Java 中的 @NotNull 和 @Nullable 注解可以帮助 kotlin 判断 Java 对象类型，其他语言类似
        val eCode = findViewById<EditText>(R.id.et_code)
        eCode.setText("hencoder")
        
        //delegate.findViewById 方式使用了 @Nullable 注解，表示该方法返回的对象可能为空，因此在 kotlin 中，该方法返回的对象类型就为可空类型 
        val code = delegate.findViewById<EditText>(R.id.et_code)
        code?.setText("hencoder")
        
        
        et_username.setText(get(usernameKey))
        et_password.setText(get(passwordKey))

        
        findViewById<Button>(R.id.btn_login).apply {
            setOnClickListener(this@MainActivity)
        }
        findViewById<CodeView>(R.id.code_view).apply {
            setOnClickListener(this@MainActivity)
        }
    }
    
    override fun onClick(v: View?) {
        if (v is CodeView) {
            v.updateCode()
        } else if(v is Button) {
            login()
        }
    }

    private fun login() {
        val username = et_username.text.toString()
        val password = et_password.text.toString()
        val code = et_code.text.toString()

        /**
         * verify() 函数确定只会再 login() 中用得到，因此可以直接将 verify 定义到 login 函数体中
         * 
         * 内部函数可以访问外部函数的成员。因此 verify(user: User) 中的 user 参数可以省略掉
         * 
         * 不过内部函数被调用的时候会生成一个额外的对象，也就是每次调用 login() 的时候都会生成一个用来调用 verify() 的对象。因此，如果 login() 会被频繁调用，就要考虑性能问题，不要老是创建出一堆垃圾对象出来
         * 内部函数调用时会创建处临时对象，因此除非定义内部函数真的能提高代码的可读性，否则不要使用
         */
        fun verify(user: User): Boolean {
            if (user.username.length < 4) {
                toast("用户名不合法")
                return false
            }
            if (user.password.length < 4) {
                toast("密码不合法")
                return false
            }
            return true
        }
        
        val user = User(username, password, code)
        if(verify(user)) {
            save(usernameKey, username)
            save(passwordKey, password)
            startActivity(Intent(this, LessonActivity::class.java))
        }
    }

    /**
     * 如果确定 verify 函数只会在 login 中使用到，那么可以将 verify 函数直接定义的 login() 函数中 
     */
    /*private fun verify(user: User): Boolean {
        *//**
         * if (user.username == null || user.username.length < 4) 等价于 user.username?.length ?: 0 < 4
         *
         * 说明：如果 user.username 为 null，那么user.username.length 就是 null，那么 user.username?.length ?: 0 就会返回 0
         *//*
        if (user.username.length < 4) {
            toast("用户名不合法")
            return false
        }
        if (user.password.length < 4) {
            toast("密码不合法")
            return false
        }
        return true
    }*/
}